Tools/scripts/tempcal_IMU.py

```diff
@@ -0,0 +1,522 @@
+            elif msg.SType == 1:
+                # gyro
+                gyr = Vector3(msg.X, msg.Y, msg.Z)
+                time = msg.TimeUS*1.0e-6
+                data.add_gyro(imu, T, time, gyr)
+
+        if msg.get_type() == 'IMU' and not args.tclr:
+            imu = msg.I
+
+            if stop_capture[imu]:
+                continue
+
+            T = msg.T
+            acc = Vector3(msg.AccX, msg.AccY, msg.AccZ)
+            gyr = Vector3(msg.GyrX, msg.GyrY, msg.GyrZ)
+
+            # invert the board orientation rotation. Corrections are in sensor frame
+            if orientation != 0:
+                acc = acc.rotate_by_inverse_id(orientation)
+                gyr = gyr.rotate_by_inverse_id(orientation)
+            if acc is None or gyr is None:
+                print("Invalid AHRS_ORIENTATION %u" % orientation)
+                sys.exit(1)
+
+            if c.enable[imu] == 1:
+                acc -= c.correction_accel(imu, T)
+                gyr -= c.correction_gyro(imu, T)
+
+            time = msg.TimeUS*1.0e-6
+            data.add_accel(imu, T, time, acc)
+            data.add_gyro (imu, T, time, gyr)
+
+    if len(data.IMUs()) == 0:
+        print("No data found")
+        sys.exit(1)
+
+    print("Loaded %u accel and %u gyro samples" % (len(data.accel[0]['T']),len(data.gyro[0]['T'])))
+
+    if not args.tclr:
+        # apply moving average filter with 2s width
+        data.Filter(2)
+
+    clog = c
+    c = Coefficients()
+
+    calfile = open(args.outfile, "w")
+
+    for imu in data.IMUs():
+        tmin = np.amin(data.accel[imu]['T'])
+        tmax = np.amax(data.accel[imu]['T'])
+
+        c.set_tmin(imu, tmin)
+        c.set_tmax(imu, tmax)
+
+        for axis in AXES:
+            if args.online:
+                fit = OnlineIMUfit()
+                trel = data.accel[imu]['T'] - TEMP_REF
+                ofs = data.accel_at_temp(imu, axis, clog.atcal[imu])
+                c.set_accel_poly(imu, axis, fit.polyfit(trel, data.accel[imu][axis] - ofs, POLY_ORDER))
+                trel = data.gyro[imu]['T'] - TEMP_REF
+                c.set_gyro_poly(imu, axis, fit.polyfit(trel, data.gyro[imu][axis], POLY_ORDER))
+            else:
+                trel = data.accel[imu]['T'] - TEMP_REF
+                ofs = data.accel_at_temp(imu, axis, clog.atcal[imu])
+                c.set_accel_poly(imu, axis, np.polyfit(trel, data.accel[imu][axis] - ofs, POLY_ORDER))
+                trel = data.gyro[imu]['T'] - TEMP_REF
+                c.set_gyro_poly(imu, axis, np.polyfit(trel, data.gyro[imu][axis], POLY_ORDER))
+
+        params = c.param_string(imu)
+        print(params)
+        calfile.write(params)
+
+    calfile.close()
+    print("Calibration written to %s" % args.outfile)
+
+    if args.no_graph:
+        return
+    fig, axs = pyplot.subplots(len(data.IMUs()), 1, sharex=True)
+
+    num_imus = len(data.IMUs())
+    if num_imus == 1:
+        axs = [axs]
+
+    for imu in data.IMUs():
```

Comment:
there's a lot of shared code between gyr and accel plotting here.

Reply:
i've tried to combine them, it is much harder than it looks